### PR TITLE
feat: implement enroll client on the device details screen [WPB-6439]

### PR DESCRIPTION
**Pull request by theswiftway**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-6439" title="WPB-6439" target="_blank">Sub-taskWPB-6439</a>  [iOS] Implement `enrollClient` on the device details screen.
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Solutions

Implement enroll client action on device details view

Needs releases with:

- [ ] GitHub link to other pull request

#### How to Test

**Steps**
1. Go to Settings
2. Tap on Devices
3. Tap on a client with not activated E2EI state
4. Tap on Get Certificate

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift [hunk 1]
```diff
@@ -130,6 +138,39 @@ final class DeviceDetailsViewActionsHandler: DeviceDetailsViewActions, Observabl
...
+        typealias E2ei = L10n.Localizable.Registration.Signin.E2ei
+        let e2eiCertificateUseCase = userSession.enrollE2eICertificate
+        guard let rootViewController = await AppDelegate.shared.window?.rootViewController else {
+            return
+        }
+        let oauthUseCase = OAuthUseCase(rootViewController: rootViewController)
+        try await e2eiCertificateUseCase?.invoke(
+            authenticate: oauthUseCase.invoke
+        )
+        userSession.reportEndToEndIdentityEnrollmentSuccess()
```
Comment (KaterinaWire) [inline wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift]: I don't think we should call `reportEndToEndIdentityEnrollmentSuccess()`.  It was relevant for the login flow.

Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift]: done

File: wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift [hunk 2]
```diff
@@ -130,6 +138,38 @@ final class DeviceDetailsViewActionsHandler: DeviceDetailsViewActions, Observabl
         }
         return fingerPrint.splitStringIntoLines(charactersPerLine: 16).uppercased()
     }
+
+    private func startE2EIdentityEnrollment() async throws {
+        typealias E2ei = L10n.Localizable.Registration.Signin.E2ei
+        let e2eiCertificateUseCase = userSession.enrollE2eICertificate
```
Comment (typfel) [inline wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift]: suggestion: Inject the `enrollE2eICertificate` like we do with the `getProteusFingerprint`

File: wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceInfoViewModel.swift [hunk 3]
```diff
@@ -117,8 +118,11 @@ final class DeviceInfoViewModel: ObservableObject {
     @MainActor
     func enrollClient() async {
         self.isActionInProgress = true
-        let certificate = await actionsHandler.enrollClient()
-        self.e2eIdentityCertificate = certificate
+        if let certificate = await actionsHandler.enrollClient() {
+            self.e2eIdentityCertificate = certificate
+        } else {
+            showEnrollmentCertificateError = true
+        }
```
Comment (typfel) [inline wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceInfoViewModel.swift]: If you let `enrollClient` throw we could we do 

```suggestion
        do {
            self.e2eIdentityCertificate = await actionsHandler.enrollClient()
        } catch {
            showEnrollmentCertificateError = true
        }
```

**Review by typfel — CHANGES_REQUESTED** [review]

File: wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceInfoViewModel.swift (hunk 3 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceInfoViewModel.swift]: done

File: wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift (hunk 2 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift]: done

**Review by typfel — APPROVED** [review]

File: wire-ios/Wire-iOS Tests/DeviceManagement/MockEnrolE2eICertificateUseCase.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+final class MockEnrolE2eICertificateUseCase: EnrollE2eICertificateUseCaseInterface {
+
+    var invokCalled: (((URL, String, String, String) async throws -> (String, String)) -> Void)?
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS Tests/DeviceManagement/MockEnrolE2eICertificateUseCase.swift]: Typo `invokeCalled`

File: wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift [hunk 5]
```diff
@@ -130,6 +141,37 @@ final class DeviceDetailsViewActionsHandler: DeviceDetailsViewActions, Observabl
...
+        let mlsGroupId = await fetchSelfConversationMLSGroupID() else {
+            return nil
+        }
+        return try await userSession.getE2eIdentityCertificates.invoke(mlsGroupId: mlsGroupId,
+                                                                clientIds: [mlsClientID]).first
+    }
+
+    @MainActor
+    private func fetchSelfConversationMLSGroupID() async -> MLSGroupID? {
+        return await contextProvider.syncContext.perform { [weak self] in
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift]: `return` can be omitted

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift [hunk 6]
```diff
@@ -128,7 +128,7 @@ final class ClientListViewController: UIViewController,
 
         self.initalizeProperties(clientsList ?? Array(ZMUser.selfUser()?.clients.filter { !$0.isSelfClient() } ?? []))
         self.clientsObserverToken = ZMUserSession.shared()?.addClientUpdateObserver(self)
-        if let user = ZMUser.selfUser(), let session = ZMUserSession.shared() {
+        if let user = ZMUser.selfUser(), let session = userSession as? ZMUserSession {
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift:131]: 👍

File: wire-ios/Wire-iOS Tests/DeviceManagement/MockContextProvider.swift [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+@testable import Wire
+import CoreData
+
+final class MockContextProvider: ContextProvider {
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS Tests/DeviceManagement/MockContextProvider.swift:23]: Can't this mock be autogenerated by Sourcery?

File: wire-ios/Wire-iOS Tests/DeviceManagement/MockEnrolE2eICertificateUseCase.swift (hunk 4 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/DeviceManagement/MockEnrolE2eICertificateUseCase.swift]: done

File: wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift (hunk 5 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/DeviceView/DeviceDetailsViewActionsHandler.swift]: done

File: wire-ios/Wire-iOS Tests/DeviceManagement/MockContextProvider.swift (hunk 7 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/DeviceManagement/MockContextProvider.swift:23]: I tried generating. It failed to generate anything.

Comment (markusfassbender) [reply wire-ios/Wire-iOS Tests/DeviceManagement/MockContextProvider.swift:23]: Did you try to find out why it doesn't generate? I mean this sounds like a bug in Sourcery or our setup..

Comment (markusfassbender) [reply wire-ios/Wire-iOS Tests/DeviceManagement/MockContextProvider.swift:23]: Okay I found the problem: the sourcery run script was completely removed from `WireDataModel`, but the config and everything still existed. So it was a bug, that no generation happened at all.
I've fixed it here for this case on this brach, hope that's fine.

**Review by markusfassbender — APPROVED** [review]

File: wire-ios/Wire-iOS Tests/DeviceManagement/MockContextProvider.swift (hunk 7 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/DeviceManagement/MockContextProvider.swift:23]: I will probably take this up later

**Comment by theswiftway** [comment]
Merging this as the failed test is just flaky

**Pull request merged**